core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -153,6 +153,7 @@ class LogManager(logDirs: Array[File],
     liveLogDirs
   }
 
+  // dir should be an absolute path
   def handleLogDirFailure(dir: String) {
```

Comment:
Taking a raw string and commenting everywhere that it has to be absolute is a bit brittle. If should probably create a `LogDir` case class that ensures that the value is always absolute.

Reply:
@ijuma I understand your point that it is a bit brittle. But I am not sure how to use case class to handle it cleanly. Do you mean we should have `case class AbsoluteLogDirPath(path: String)` and refactor methods to be e.g `def handleLogDirFailure(dir: AbsoluteLogDirPath)`? I question is when should we convert a string path into this case class? Do you recommend to convert string to this case class at the very beginning (e.g. when we receive it from user's requests and in the broker configuration) and pass it everywhere in the code where log directory is used?

Reply:
Yeah, that's the idea. I think it's fine to call it `LogDir` btw and document that it always holds an absolute path. We'd validate and convert during construction. Is this feasible? If it's a big change, we could file a separate JIRA for it.

Reply:
@ijuma It is certainly feasible. I don't mind making all the code changes for this. My concern is that we need to change the type of log dir from String to this case class in a few places (e.g. every time we read KafkaConfig.logDirs, logDir from DescribeLogDirRequest and logDir from AlterReplicaDirRequest) and other developer may come up with different opinion on the resulting code later because code style preference can be subjective.

Could I create a separate pull request for this? If so, I will submit a patch for it today and you can review it whenever you are free. If no, I will make this change in this class and hopefully you can get time to review sooner :)

Reply:
@ijuma Can you take a look at https://github.com/apache/kafka/pull/3763 and see if it address your comment? If you prefer I can close that pull request and move the change here.

I still name the case class because it makes it more explicit that this case class should wrap around a string that represents an absolute log directory path. Ideally developer don't need to read the comment to understand what this class is doing. What do you think?

Reply:
Thanks for submitting the other PR. I'm OK with @becketqin merging this one if he's happy with it. And the second PR is then a refactoring to make the code less brittle.

Reply:
@ijuma Thanks for the review. I'll merge this one as it is a quick bug fix.